File: android/app/src/main/java/com/turntable/barcodescanner/TurnTableSearchDeepLink.kt
package com.turntable.barcodescanner

import android.net.Uri

/** Parses `turntable://search?barcode=&artist=&album=` deep links. */
object TurnTableSearchDeepLink {
    const val SCHEME = "turntable"
    const val HOST_SEARCH = "search"

    data class Parsed(
        val barcode: String,
        val artist: String,
        val album: String,
        /** True when artist and/or album arrived from the scanner (skip barcode prefetch). */
        val hasResolvedRelease: Boolean,
    ) {
        val secondaryTerms: String
            get() = when {
                artist.isNotEmpty() && album.isNotEmpty() -> "$artist - $album"
                album.isNotEmpty() -> album
                artist.isNotEmpty() -> artist
                else -> ""
            }
    }

    fun parse(uri: Uri?): Parsed? {
        if (uri == null) return null
        if (uri.scheme != SCHEME) return null
        if (uri.host != HOST_SEARCH) return null
        val barcode = uri.getQueryParameter("barcode")?.trim().orEmpty()
        val artist = uri.getQueryParameter("artist")?.trim().orEmpty()
        val album = uri.getQueryParameter("album")?.trim().orEmpty()
        if (barcode.isEmpty() && artist.isEmpty() && album.isEmpty()) return null
        return Parsed(
            barcode = barcode,
            artist = artist,
            album = album,
            hasResolvedRelease = artist.isNotEmpty() || album.isNotEmpty(),
        )
    }
}

File: ios/turnTable/TurnTableSearchDeepLink.swift
import Foundation

/// Parses `turntable://search?barcode=&artist=&album=` from the Scanner companion.
enum TurnTableSearchDeepLink {
    struct Parsed {
        var barcode: String
        var artist: String
        var album: String
        var hasResolvedRelease: Bool {
            !artist.isEmpty || !album.isEmpty
        }
        var secondaryTerms: String {
            switch (artist.isEmpty, album.isEmpty) {
            case (false, false): return "\(artist) - \(album)"
            case (true, false): return album
            case (false, true): return artist
            default: return ""
            }
        }
    }

    static func parse(_ url: URL?) -> Parsed? {
        guard let url = url,
              url.scheme?.lowercased() == "turntable",
              url.host?.lowercased() == "search" else { return nil }
        let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        func q(_ name: String) -> String {
            items.first { $0.name == name }?.value?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        }
        let barcode = q("barcode")
        let artist = q("artist")
        let album = q("album")
        if barcode.isEmpty && artist.isEmpty && album.isEmpty { return nil }
        return Parsed(barcode: barcode, artist: artist, album: album)
    }
}
